alibi_detect/utils/tensorflow/misc.py

```diff
@@ -95,3 +96,40 @@ def clone_model(model: tf.keras.Model) -> tf.keras.Model:
         except NotImplementedError:
             config = {}
         return model.__class__.from_config(config)
+
+
+def check_model(model: tf.keras.Model, raise_error: bool = False) -> None:
+    """
+
+    Parameters
+    ----------
+    model
+    raise_error
+
+    Raises
+    ------
+
+    """
+    # Message to display if there are problems
+    msg = "The TensorFlow model may have been loaded incorrectly. This could be because `get_config` and/or " \
+          "`from_config` methods were defined incorrectly. Otherwise, it could be because custom objects have not " \
+          "been provided. For more guidance see the TensorFlow tab at " \
+          "https://docs.seldon.io/projects/alibi-detect/en/stable/overview/saving.html#supported-ml-models."
+
+    try:
+        # Check if model is a `RevivedNetwork` rather than the real original model (this occurs when subclassed models
+        # are loaded without all custom objects being provided
+        # Note, could also do `if model.__class__.__base__.__name__ == 'RevivedNetwork':`
+        if model.__class__.__module__ == 'keras.saving.saved_model.load':
+            # Raise error (this will be caught and re-raised below)
+            raise ValueError('The model appears to be a `keras.saving.saved_model.load.RevivedNetwork. This suggests '
+                             'a subclassed model has been loaded without all custom objects being provided.')
+
+        # Check model cloning doesn't raise error
+        clone_model(model)
+
+    except Exception as error:
+        if raise_error:
+            raise ValueError(msg) from error
+        else:
+            warnings.warn(msg + f"Original error message: \n\t{error}")
```

Comment:
Not sure this works that well together, e.g. we check if cloning works but it any case, the error is never raised. So effectively this would blow up again when using inside detectors that do use cloning? Or is the intention to call this method there also?

Reply:
The problem I am having here is that there are a large number of possible failure modes that can occur due to not specifying the `custom_objects` properly. Depending on the exact tf version, and whether the model itself is subclassed, or just layers, either `ValueError`, `TypeError`'s or `NotImeplemented` errors are raised, and this is either during inference or cloning. 

The widest net I've been able to cast is to simply test if cloning works, and to check if there might be problems at inference, check whether the model is a `RevivedNetwork`, which is what it's loaded as if custom objects are missing (I don't try to actually call the model as data isn't available at this point). 

A wide net is nice in term of hopefully catching most errors, but has the downside of throwing errors when things might have actually worked. For example, the `RevivedNetwork`'s generally work for inference but not cloning. So a user can actually get away with not passing `custom_objects` if the model is just to be used for preprocessing. Hence why I went with a warning instead of error...

Maybe one compromise is to raise a warning if the model is a `RevivedNetwork`, since this can cause issues, but raise the `ValueError` (with the more coherent message) if cloning fails?

Reply:
I'm not sure, cloning failure is only relevant for a specific set of detectors, so somehow feels like it should be checked only for that subset. Maybe leave this as is and wrap the relevant detector calls to `clone()` in `try/except` with a customized Alibi error message?

Reply:
Sorry, just coming back on this, I've thought about it a bit more. There are just so many different failure modes, and the problem with just checking if cloneable is that there is one nasty failure mode where inference works, cloning works, BUT inference after cloning doesn't work 🤯 (this occurs in example 5 [here](https://colab.research.google.com/gist/ascillitoe/bab98843256d99be14ed6fa4a85ff640/tensorflow_save-load.ipynb), the custom `call` method is lost when cloning...). 

This is what motivates casting the wide net by checking if a `RevivedNetwork` too. But then throwing a warning for this would prevent users getting away with cases that would otherwise work with no `custom_objects` provided (generally when no cloning). 

I see two options:
1. Leave pretty much as is. Downside at the moment is by swallowing the `ValueError` from cloning and raising as a warning, it can be somewhat lost in the later actual errors that occur (i.e. when cloning down for real). 
2. Turn the warning into an error. Be up front about requiring `custom_objects` to be passed in all cases where there are custom objects involved. This is slightly less convenient for the user, but means we don't have to worry about these many different failure modes that might also change in future versions.

Reply:
Yes exactly. It would mean any models containing custom classes loaded without providing those custom classes (via registering, or `custom_objects` kwarg) would be pretty much guaranteed to fail. Thinking about it more this would probably be what I would lean towards. Although to be clear, this does also mean that, for example, if a user had a relatively simple `tf.keras.Sequential`, but with one custom layer, they would have to provide this layer at load time even if they don't use the model in a detector with any cloning...

Reply:
The `tensorflow` behaviour with this has been so fluid in versions 2.9, 2.10 and 2.11 that I do think it'd be the safest option though... (for now)